Feature/echo 508 leaf only option

tgoff opened the pull request.

tgoff committed:
ECHO-508 Add simplify_output attribute to tree control

This new attribute will cause tree output to be filtered to include only leaf
nodes and parents with all descendants selected

tgoff committed:
ECHO-508 Add simplify_output attribute to tree control

This new attribute will cause tree output to be filtered to include only leaf
nodes and parents with all descendants selected

ECHO-508 Add tests

tgoff committed:
ECHO-508 Add tests

tgoff committed:
Merge branch 'feature/ECHO-508_LeafOnlyOption' of https://github.com/nasa/edsc-echoforms into feature/ECHO-508_LeafOnlyOption

bilts commented on spec/src/features/tree-spec.coffee:
```diff
@@ -282,6 +282,179 @@ describe '"tree" control', ->
+      expect(form.echoforms('serialize')).toMatch(/>level_3_child_2</)
+      expect(form.echoforms('serialize')).not.toMatch(/>level_2_child_2</)
+
+    it "defaults to true", ->
```
Parent contexts should use `describe`, not `it`, `it` is for the actual specs.

bilts commented on spec/src/features/tree-spec.coffee:
```diff
@@ -282,6 +282,179 @@ describe '"tree" control', ->
+      expect(form.echoforms('serialize')).not.toMatch(/>level_2_child_2</)
+
+    it "defaults to true", ->
+      form = template.form(dom, ui: ui, model: model)
```
I assume these three lines should be (and similar elsewhere in this file) should be `beforeEach` blocks.

bilts commented on spec/src/features/tree-spec.coffee:
```diff
@@ -282,6 +282,179 @@ describe '"tree" control', ->
+
+
 
   describe "other test cases", ->
```
I know you didn't change this, but could you? :)

tgoff replied:
which lines are you referring to?

bilts commented on src/controls/extras/tree.coffee:
```diff
@@ -79,16 +80,29 @@ class Tree extends Typed
+    required = @inputs().find('li[item-required=true][node_value][item-relevant= true]')
+    checked_required_nodes = checked.add(required)
+
+    if @simplify_output
```
Does this work bidirectionally? i.e. if I give you a model pre-populated with the simplified output, can you reverse it into tree selections? I'm fairly certain the plugin itself needs that to work, and EDSC definitely depends on it working.

tgoff replied:
do you mean does specifying a parent node in the initial model cause all its children to be selected?  if so then yes it does.  note that simplified output is not formatted any differently, it is just a subset of the values which would be output without simplified set to true.  The lines you commented on are actually not new functionality, just a refactoring of the existing code.  then I add new logic to select out a subset of those values if simplify is enabled.

bilts replied:
Great, thanks

tgoff committed:
ECHO-508 do not allow nodes to be collapsed

Collapsing nodes is disabled as we need them to be open so they
remain in the DOM and are available for the jquery selectors

Also refactor some tests
